Test Case 7

Aill3n opened the pull request:
For this exercise, I tried to practice and apply aspects of modern Java and functional programming, which would go well with my EnumMap approach. 
I used streams, maps, lambda filters, method references and Optional.

Tests:
From the home page go to the shop page
Buy the first product you find with a given price. For example buy a product with price 9.99
Validate that the cart menu displays 1

Test case design
Coding with intent
Model repeating elements
Dependency injection for root element
Use of proper types for currency
Use of strategy pattern  to solve code repetition
Selenium API:
Find elements within elements

Aill3n committed:
Adding tests for tc7

Aill3n committed:
minor adjustments

airvinenz reviewed (changes requested):
Again, nicely written.
Clean, tidy, readable code.

The wonderful characteristic of your solution so far has been "pragmatic".
You're only building what you need, and it is actually GREAT.

The problem we have here is that when you buy a "10.99 item", you're actually selecting an index from your enummap with that price, instead of from the DOM -- these might not always match.

This is also true for your test selecting by name.

So your tests are breaking rule 2:
"an automated test must do what it says" :-)

Aill3n committed:
massive refactor replacing Enum logic

Aill3n committed:
renaming method

Aill3n commented:
Thank you for the review, Alex, and for showing me alternative ways to solve the problem 🙇‍♀️.

I have opted to refactor my entire shop logic and adopt something similar to what you presented to me last week.

It would be possible to keep my original design and modify the logic for the locators, but I wanted to challenge myself with this refactoring. 
I started it in old Java, and after my new logic was in place, I modified the new methods to use streams instead of creating multiple lists and looping through them.

Let's see what you think ☕.

Aill3n committed:
changes from code review with alex

Aill3n committed:
Addressing code review

Aill3n commented:
Changes done:

- Moved all the page models from the "pages" package to a new package called "model".
- Added the Product model to the "model" package.
- Improved the Product class to have a button WebElement and added an addToCart method.
- Removed repetitive code from the ShopPage.
- Improved validation messages in the ShopTest file.
- Added a new class for CartPage that will be used in the future.

Aill3n committed:
improving assertion messages

airvinenz commented on src/test/java/com/planittesting/cloud/jupiter/model/ShopPage.java:
```diff
@@ -0,0 +1,62 @@
+                .collect(Collectors.toList());
+    }
+
+    public Optional<BigDecimal> getProductPriceByName(String productName) {
```
This is an overcomplication.
you already have findFirstProductByPrice(...) which returns a product.
Since you then have a Product, you should be able to use something like `myProduct.getPrice()` in your test.

airvinenz commented on src/test/java/com/planittesting/cloud/jupiter/tests/ShopPageTest.java:
```diff
@@ -1,22 +1,45 @@
 
     @Test
-    public void validateItemPrice() {
+    public void validateItemPriceTest() {
```
This can be much simpler:

    public void validateItemPriceTest() {
        // Step 1: From the home page go to the shop page
        ShopPage shopPage = basePage.openShopPage();

        // Step 2: Validate a given price for a given product title. For example given Teddy Bear validate that the price is 12.99
        Product bunny = shopPage.getProductByName("Fluffy Bunny");
        assertEquals(9.99, bunny.getPrice(), "Bunny Price Check");

        <snip>
}

airvinenz commented on src/test/java/com/planittesting/cloud/jupiter/tests/ShopPageTest.java:
```diff
@@ -1,22 +1,45 @@
+        Optional<Product> product = shopPage.findFirstProductByPrice(price);
+        assertTrue(product.isPresent(), "Validating product with price $14.99 is on the shop page");
+
+        product.get().clickBuyButton();
```
if we have a product here, why do we need to go a get before we can click 'buy' ?

Aill3n replied:
Because this "product" is the Optional<product> container object. To access the product within the optional container, I need to get it with get().

airvinenz commented on src/test/java/com/planittesting/cloud/jupiter/tests/ShopPageTest.java:
```diff
@@ -1,22 +1,45 @@
-        assertEquals(expectedValue, actualValue, "Validating price found for product: " + Toy.FLUFFY_BUNNY.name());
+        // Step 3: Validate that the cart menu displays 1
+        Integer actualQuantityInCart = shopPage.getCartItemCount();
+        assertEquals(1,actualQuantityInCart, "Validating the cart contains 1 item after adding a product.");
```
too many words!
the message can be as simple as "Cart Count".  The assertion will do the rest for you

...and then we can have a conversation about whether this is even a good test or not.
I would suggest it's naive (maybe deliberately!).  Just because we clicked 'buy' on one product doesn't mean the counter shoudl say "1".  What would be better? :-)

Aill3n replied:
It's quite long, yeah 😅 . I can change it to "Cart count."

I can access the Cart page and make sure the product is present there, in addition to checking that the count is 1. Is this what is expected?

airvinenz replied:
This test is only about checking that the counter in the top corner increments :-)

Aill3n committed:
addressing code review, comments 1, 2 and 3

Aill3n commented:
Addressed comments 1, 2 and 3. Will work on comment 4 to add an incrementation check to the cartCount. 🔥

Aill3n committed:
addresing naive cartCount

Aill3n commented:
Addressed comment 4. 
Ready for review again, thanks for your time, Alex 🙏

Aill3n commented on src/test/java/com/planittesting/cloud/jupiter/tests/ShopPageTest.java:
```diff
@@ -17,9 +17,9 @@ public void validateItemPriceTest() {
-        Product product = shopPage.getProductByPrice(price);
-        assertEquals("Fluffy Bunny", product.getName(), "Name match.");
-        assertEquals(price, product.getPrice(), "Product price.");
+        Product fluffyBunny = shopPage.getProductByPrice(price);
```
I ctrl+z too much 😅

Aill3n committed:
fixing variable name

airvinenz commented on src/test/java/com/planittesting/cloud/jupiter/tests/ShopPageTest.java:
```diff
@@ -18,28 +16,26 @@ public void validateItemPriceTest() {
-        assertTrue(actualValue.isPresent(),"Confirming product 'Fluffy Bunny' is present in the shop page.");
-        assertEquals(expectedValue, actualValue.get(),"Validating the price " + expectedValue + " for Fluffy Bunny is correct.");
+        BigDecimal price = new BigDecimal("9.99");
+        Product product = shopPage.getProductByPrice(price);
```
this one needs to get the product by name :-)  " For example given Teddy Bear validate that the price is ..."

Aill3n replied:
Hoping I could fix it properly this time. 😅

airvinenz commented on src/test/java/com/planittesting/cloud/jupiter/tests/ShopPageTest.java:
```diff
@@ -18,28 +16,26 @@ public void validateItemPriceTest() {
-        BigDecimal price = new BigDecimal("14.99");
-        Optional<Product> product = shopPage.findFirstProductByPrice(price);
-        assertTrue(product.isPresent(), "Validating product with price $14.99 is on the shop page");
+        Integer currentCartCount = shopPage.getCartItemCount();
```
If you read this one out loud, it doesn't correctly "say what it does" on the last line.
The variable name should be "initialCartCount" :-)

airvinenz reviewed (changes requested):
So close!

Aill3n committed:
adding lambda and changing variable naming

airvinenz reviewed (approved):
Yes!
